### DISCUSSION TITLE
Using github app vs actions for site and data builds

### DISCUSSION THREAD

**Discussion by elray1** [hubverse-org/hub-dashboard-predtimechart#34] [General]
I am starting a discussion thread to re-surface some questions that have come up in Slack threads about our setup with building dashboards with a github app vs using github actions.  In part, this is about nailing down our scope of work for this dashboard tool, and in part it's about making sure we are using the best approach to meet the work that's in scope.  To try to sum up, I think there are two main considerations:

1. How to perform scheduled and ad-hoc end-of-round data and site builds in a per-hub fashion.
2. Whether or how to handle custom website content that is programmatically generated.

I don't necessarily think either of these rules out using the github app, but I don't have a clear vision of how to handle these things using the app and I have the sense that doing so may be more complex using the app than it would be using github actions.  That said, maybe there are solutions I'm not thinking of. And on the other hand, it may also be that the right answer is just that we don't support one or both of these. I'll write more about each of these below.  I'm more worried about point 1 than point 2.

## 1. end-of-round builds

I think there are ~3 or 4 separate things that we want to be able to do at the end of each modeling round in order to support a hub's dashboard:

(a) potentially, rebuild static site content [see consideration 2 below, programmatically generated content]
(b) build target data for the viz
(c) build model output data for the viz
(d) [eventually, not yet supported] build score data for an evaluation page.

In principle, if the world was well aligned, these things would happen only once per round in a predictable way.  But in practice, weird things happen.  For example, last week for flu the data release on Wednesday had issues and they ended up releasing corrected data on Thursday and allowing (new or revised) forecast submissions up till noon on Friday.  So for that hub, the correct data vintage to use for reference date 2024-11-23 would have been Thursday, and the correct model outputs to use would have been the ones that were in the FluSight repo as of Friday afternoon.  Of note, using the data published by CDC on Friday afternoon would not have been correct.  However, the same data source was used for the covid hub, and corrected data for that hub also came out on Thursday -- but the covid hub administrators decided not to re-open forecast submissions.  The correct data vintage for the covid hub was the data from Wednesday, updating to the revised data would have been wrong.

This example illustrates a few challenges:

- for a viz, ideally we would show the latest target data as of the effective submission date.  we don't want to always rebuild/replace the target data every day up until the reference date
- manual overrides may be necessary
- you may want to rebuild your model output data but not your target data
- these things can be different for different hubs

It's much easier for me to think about how to deal with this if there are separate processes for triggering target data and model output data builds that can be scheduled and triggered on a per-hub basis.  I'm very much open to other answers, but it does feel like our current setup leads to some awkardness here.

## 2. programmatically generated content

Here are examples of situations where we have said "it would be nice if this hub could build some of its dashboard content programmatically":

- one-off evaluation pages for flusight and covid hub.  We're hoping to eventually build a "real" eval tool, but in the short term the main option we've thought of is to build a Rmd or qmd file that does evaluations each week.
- visualizations of forecasts that are not handled already by predtimechart
- reports for the variant nowcast hub describing what the predictions are saying.

I want to note that all of this is a change in scope from what I first outlined when I described the vision for this dashboard tool to Matt & Zhian a couple of months ago.  At that time, Zhian specifically asked, "are hub admins going to want to be able to run R code?" and my answer was "maybe, but it's out of scope".  My answer now is that it still may be reasonable to say this is out of scope, but it would be nice if we could support it.

I think there are a couple of potential challenges here: (a) managing dependencies for the code; (b) similar to consideration 1 above, scheduling builds of this hub-specific content.  I wonder if a way to deal with these might be to say that hubs that want this are responsible for setting up a separate build process for this content, that produces static site content in a markdown file? For example, they could set up their own github action to build this static content to a .md file, which would then be handed off to the rest of our dashboard build process?

**Comment by zkamvar** [comment]
The app vs actions dichotomy is a bit of a false one and it's not so useful to think about this in the scope of the questions posed here. 

The questions posed here are more about the processes and how each one can be separated. The separation of these processes is the same regardless of whether or not we use an app to centrally run things or a series of distributed GitHub workflows to run them. 

## Process

Right now, the workflow runs once per day and on demand. For each installed repository, it does the following in parallel runs:

1. builds the website and pushes it to `gh-pages`
2. builds the target and forecast data and pushes it to `ptc/data`

It's built in this way because I wanted to make sure the process of building and then pushing worked before I messed with the finer-grained triggering and filtering of processes. 

### Website

This is by far the least resource-intensive part of the process. For this we run a docker container that contains the website template inside the dashboard repository and push the result to `gh-pages`. This usually completes in about a minute for each repo

### Data

This requires us to fetch a few resources:

1. install `hub-dashboard-predtimechart`
2. the `ptc/data` branch from the dashboard repository
3. the hub repository

At the moment, we are building both the target and forecast data in the same workflow, but it's not difficult to separate these since they are different concerns. They are currently built together because we've set them to live in the same branch (but again, not difficult to separate them. They don't touch each other's file paths, so `git pull` with rebase before a `git push` will help resolve any conflicts. 

## Deployment

The question of workflow vs app is a question of deployment: do we want to have a centrally-managed workflow (app) or do we want a distributed workflow (similar to what we have now for the validations)? And note that these are not necessarily mutually exclusive (an app can trigger a workflow on a repository). 

I went with a centrally-managed workflow because it reduces the burden on the users and makes it easier for me to iterate quickly. We've already gotten feedback from the FluSight folks that they do not have the capacity or authorization to manage yet another repository or website. I am also _deeply_ familiar with some of the pitfalls of having distributed workflows that you attempt to centrally manage. One pitfall is that GitHub workflow files are complex and [understanding them is above most hubAdmins' pay grade](https://github.com/swcarpentry/r-novice-gapminder/pull/806#issuecomment-1287326865). This leads to people not updating their workflows and things eventually becoming deprecated and breaking.
